### ISSUE TITLE
Changes to i8data for DatetimeIndex

### ISSUE THREAD

**Issue by TomAugspurger** [pandas-dev/pandas#24559]
Master currently has an (undocumented) (maybe-) API-breaking change from 0.23.4 when passed integer values

0.23.4

```python
In [2]: i8data = np.arange(5) * 3600 * 10**9

In [3]: pd.DatetimeIndex(i8data, tz="US/Central")
Out[3]:
DatetimeIndex(['1970-01-01 00:00:00-06:00', '1970-01-01 01:00:00-06:00',
               '1970-01-01 02:00:00-06:00', '1970-01-01 03:00:00-06:00',
               '1970-01-01 04:00:00-06:00'],
              dtype='datetime64[ns, US/Central]', freq=None)
```

Master

```python
In [3]: pd.DatetimeIndex(i8data, tz="US/Central")
Out[3]:
DatetimeIndex(['1969-12-31 18:00:00-06:00', '1969-12-31 19:00:00-06:00',
               '1969-12-31 20:00:00-06:00', '1969-12-31 21:00:00-06:00',
               '1969-12-31 22:00:00-06:00'],
              dtype='datetime64[ns, US/Central]', freq=None)

```

---

Attempt to explain the behavior: In 0.23.4, passing an `ndarray[i8]` was equivalent to passing `data.view("M8[ns]")`

```python
# 0.23.4
In [4]: pd.DatetimeIndex(i8data.view("M8[ns]"), tz="US/Central")
Out[4]:
DatetimeIndex(['1970-01-01 00:00:00-06:00', '1970-01-01 01:00:00-06:00',
               '1970-01-01 02:00:00-06:00', '1970-01-01 03:00:00-06:00',
               '1970-01-01 04:00:00-06:00'],
              dtype='datetime64[ns, US/Central]', freq=None)
```

On master, integer values are treated as unix timestamps, while M8[ns] values are treated as wall-times in the given timezone.

```python
# master
In [4]: pd.DatetimeIndex(i8data.view("M8[ns]"), tz="US/Central")
Out[4]:
DatetimeIndex(['1970-01-01 00:00:00-06:00', '1970-01-01 01:00:00-06:00',
               '1970-01-01 02:00:00-06:00', '1970-01-01 03:00:00-06:00',
               '1970-01-01 04:00:00-06:00'],
              dtype='datetime64[ns, US/Central]', freq=None)
```

---

**Reason for the change**

There are four cases of interest:

```
In [4]: arr = np.arange(5) * 24 * 3600 * 10**9
In [5]: tz = 'US/Pacific'

In [6]: a = pd.DatetimeIndex(arr, tz=tz)
In [7]: b = pd.DatetimeIndex(arr.view('M8[ns]'), tz=tz)
In [8]: c = pd.DatetimeIndex._simple_new(arr, tz=tz)
In [9]: d = pd.DatetimeIndex._simple_new(arr.view('M8[ns]'), tz=tz)

In [10]: a
Out[10]: 
DatetimeIndex(['1970-01-01 00:00:00-08:00', '1970-01-02 00:00:00-08:00',
               '1970-01-03 00:00:00-08:00', '1970-01-04 00:00:00-08:00',
               '1970-01-05 00:00:00-08:00'],
              dtype='datetime64[ns, US/Pacific]', freq=None)

In [11]: b
Out[11]: 
DatetimeIndex(['1970-01-01 00:00:00-08:00', '1970-01-02 00:00:00-08:00',
               '1970-01-03 00:00:00-08:00', '1970-01-04 00:00:00-08:00',
               '1970-01-05 00:00:00-08:00'],
              dtype='datetime64[ns, US/Pacific]', freq=None)

In [12]: c
Out[12]: 
DatetimeIndex(['1969-12-31', '1970-01-01', '1970-01-02', '1970-01-03',
               '1970-01-04'],
              dtype='datetime64[ns, US/Pacific]', freq=None)

In [13]: d
Out[13]: 
DatetimeIndex(['1969-12-31', '1970-01-01', '1970-01-02', '1970-01-03',
               '1970-01-04'],
              dtype='datetime64[ns, US/Pacific]', freq=None)
```

In 0.23.4 we have `a.equals(b)` and `c.equals(d)` but no way to pass data in a way that was constructor-neutral.  In master we now have `a` match `c` and `d`.  At some point in the refactoring process we changed that, but off the top of my head I don't remember when or if this was the precise motivation or just a side-benefit.

BTW _simple_new was also way too much:

```
        if getattr(values, 'dtype', None) is None:
            # empty, but with dtype compat
            if values is None:
                values = np.empty(0, dtype=_NS_DTYPE)
                return cls(values, name=name, freq=freq, tz=tz,
                           dtype=dtype, **kwargs)
            values = np.array(values, copy=False)

        if is_object_dtype(values):
            return cls(values, name=name, freq=freq, tz=tz,
                       dtype=dtype, **kwargs).values
        elif not is_datetime64_dtype(values):
            values = _ensure_int64(values).view(_NS_DTYPE)
```

---

**Was this documented**?

http://pandas.pydata.org/pandas-docs/stable/generated/pandas.DatetimeIndex.html mentions that it's "represented internally as int64".

The (imprecise) type on `data` is "Optional datetime-like data"

I don't see anything in http://pandas.pydata.org/pandas-docs/stable/timeseries.html suggesting that integers can be passed to DatetimeIndex.

**Comment by TomAugspurger** [comment]
@jbrockmendel could you fill in the "Reason for the change" section? IIUC, it was to simplify `DatetimeIndex._simple_new`?

**Comment by mroeschke** [comment]
Possibly related, I worked on this topic in https://github.com/pandas-dev/pandas/pull/21216. Rational was that `DatetimeIndex(ints, tz=tz)` should behave similarly as `Timestamp(int, tz=tz)`, and ints cannot necessarily represent wall times. 

I have an open issue https://github.com/pandas-dev/pandas/issues/20257 to document passing integers into `Timestamp` which, similarly to `DatetimeIndex`, will treat integers as epoch (unix) timestamps

**Comment by jbrockmendel** [comment]
@mroeschke thanks. IIRC consistency with to_datetime was also a consideration

**Comment by jorisvandenbossche** [comment]
> IIRC consistency with to_datetime was also a consideration

`to_datetime` has no `tz` keyword, only `utc=True`, but for UTC wall time or unix epochs are the same?

**Comment by jorisvandenbossche** [comment]
The `Timestamp` / `DatetimeIndex` consistency is a good reason to change one of both (since they are inconsistent in 0.23.4). 
And I think the unix epoch way makes sense for integer input. Although I think it can also be confusing that `DatetimeIndex(int_array, ..)` and `DatetimeIndex(int_array.view('M8[ns]), ..)` give different results.

Regarding"was this documented": it might not have been documented clearly, but it is long standing behaviour: either we think people don't use it / shouldn't use it (but then why bother changing it? Shouldn't we then rather deprecate the whole ability of passing integer data, instead of changing the behaviour?), or either people do use it, and this change will break that usage. 
Of course, a break for a limited number of people might be worth the trade-off for a big win within our code base. But is this consistency between `__new__` and `_simple_new` that important code-technical?

**Comment by jorisvandenbossche** [comment]
Any other thoughts / replies here?

**Comment by jbrockmendel** [comment]
I think the overriding internal-consistency concern is the one @mroeschke reminded us of: this should behave like Timestamp constructor.

@TomAugspurger would "fixing" this behavior be difficult?  Last time I tried something similar I got test_packers failures that I couldn't figure out.

**Comment by TomAugspurger** [comment]
Demo of the inconsistency on 0.23.4

```python
In [9]: i8 = pd.Timestamp('2000', tz='CET').value

In [10]: pd.Timestamp(i8)
Out[10]: Timestamp('1999-12-31 23:00:00')

In [11]: pd.Timestamp(i8, tz="CET")
Out[11]: Timestamp('2000-01-01 00:00:00+0100', tz='CET')

In [12]: pd.DatetimeIndex(np.array([i8]))
Out[12]: DatetimeIndex(['1999-12-31 23:00:00'], dtype='datetime64[ns]', freq=None)

In [13]: pd.DatetimeIndex(np.array([i8]), tz="CET")
Out[13]: DatetimeIndex(['1999-12-31 23:00:00+01:00'], dtype='datetime64[ns, CET]', freq=None)
```

On master, Out[13] matches Out[11].

That consistency is certainly worth striving for.

---

If we wanted a graceful deprecation warning, then the DTI constructor would

1. check for i8data & tz
2. tell users to... what? What recourse would they have? If they want wall times (the previous behavior) then they can `.view("dateime64[ns]")` and pass that. If they want unix epochs (the future behavior) then they should ...?
3. Update all of pandas to use the recommended behavior from 2 (which is TBD).

Putting that warning in place would give us an idea of how difficult this would be to change. I'll see what turns up. Since `_simple_new` already does the "right" (future) thing, it may not be too bad...

**Comment by jorisvandenbossche** [comment]
> If they want unix epochs (the future behavior) then they should ...?

`to_datetime` ? (if it would finally have a `tz` option)

The other option could also be to deprecate passing integers, if we find it too confusing what it should result in.

**Comment by mroeschke** [comment]
Since `tz` effectively means `tz_localize`, in an ideal world I would only want integer data to only accept `tz='UTC'/None`. I think `Timestamp(i8, tz=tz)`'s localization to UTC then conversion to tz is a little too magic although it's the more appropriate behavior. Therefore I had thought the `DatetimeIndex(i8_array, tz=tz)` behavior was a bug, but since our policy wasn't set in documentation deprecation is appropriate.

For 2, if a user wants epoch timestamps (future behavior) they should use `DatetimeIndex(i8_array, tz='UTC').tz_convert(tz)` (or even `to_datetime` since it even has an `origin` parameter). I have a feeling we may rely on passing i8 into these constructors internally so deprecating the behavior may warrant a lot of refactoring down the line?

**Comment by jbrockmendel** [comment]
#24623 is now passing.  PTAL.

**Comment by jorisvandenbossche** [comment]
@jbrockmendel it is not fully clear to me what #24623 exactly does related to the discussion above. Can you summarize it here? 
(e.g. does it change back the behaviour of DatimeIndex? (but I thought you were opposed to that?) And I don't think it adds a deprecation warning? ..)

**Comment by jbrockmendel** [comment]
> it is not fully clear to me what #24623 exactly does related to the discussion above. Can you summarize it here?

There are two central changes in #24623, both of which were the status quo before #24024.

1) make `DatetimeArray.__init__` behave like `DatetimeIndex.__new__` and `Timestamp.__new__` by treating int64 values as unix timestamps and datetime64 as wall-times.

2) Do frequency validation in `DatetimeArray.__init__`, preventing users from creating invalid instances with the public constructor.

**Comment by jorisvandenbossche** [comment]
Then how is that related to this issue? This issue is about the change in behaviour of handling integer values in the DatetimeIndex constructor, not about the DatetimeArray constructor

**Comment by jbrockmendel** [comment]
> Then how is that related to this issue? 

If it isn't closely enough related for your tastes then feel free to ignore.

**Comment by jorisvandenbossche** [comment]
You wrote the PR, I think you are better placed to judge to what extent it is related to the discussion here ;) 
It's just because you mentioned it here as if it is resolving this issue (but maybe I misunderstood that) that I am asking for clarification.

**Comment by jbrockmendel** [comment]
I'm actually currently getting side-tracked by trying to figure out exactly what is and isn't internally-consistent.  It looks like 24623 makes i8 behavior consistent between DTA/DTI/Timestamp and M8[ns] behavior consistent between DTA/DTI, but breaks the existing consistency in master between DTA/Timestamp.

I'll try to put together a grid for these...

**Comment by jorisvandenbossche** [comment]
And an answer to the comment of @mroeschke 

> Since tz effectively means tz_localize, in an ideal world I would only want integer data to only accept tz='UTC'/None.

That sounds as a logical rule to me.

**Comment by jbrockmendel** [comment]
> Since tz effectively means tz_localize, in an ideal world I would only want integer data to only accept tz='UTC'/None.

I'm not sure I understand this rule.  Is the idea that the constructor would only be able to create tz-aware {n \choose k}(DTA/DTI/Timestamp) with UTC and them make the user do tz_convert as another step?

**Comment by mroeschke** [comment]
In an _ideal_ world I would prefer that since it's more explicit, but I think it would require deprecating `Timestamp` behavior with ints and tz.

I would be fine just propagating the current `Timestamp` behavior and more clearly documenting interger with `tz` arg behavior across the board since I think we would have do deprecate less?

**Comment by jbrockmendel** [comment]
I've made things more difficult here by voicing opinions while being misinformed.  Apologies all around.

Recap/Double-Checking behavior from 0.23.4 and master.  In all cases passing tz/dtype corresponding to US/Pacific.

                            
```
ts = pd.Timestamp.now('US/Pacific').round('s')
dti = pd.DatetimeIndex([ts])
tz = dti.tz
dtype = dti.dtype

results = [
    pd.Timestamp(np.int64(ts.value), tz=ts.tz),
    pd.Timestamp(np.datetime64(ts.value, "ns"), tz=ts.tz),
    pd.Timestamp(float(ts.value), tz=ts.tz),
]
assert all(x == ts for x in results)   # <-- holds in both 0.23.4 and in master

args = [
    dti.asi8,
    dti.asi8.view('datetime64[ns]'),
    dti.asi8.astype(np.float64),
    dti.asi8.astype(object),
    dti.asi8.view('datetime64[ns]').astype(object),
    dti.asi8.astype(np.float64).astype(object)
]

news = [pd.DatetimeIndex(x, dtype=dtype) for x in args]
simples = [pd.DatetimeIndex._simple_new(x, tz=tz)
           for x in args if x.dtype == 'i8' or x.dtype == 'M8[ns]']
inits = [pd.core.arrays.DatetimeArray(x, dtype=dtype)
         for x in args if x.dtype == 'i8' or x.dtype == 'M8[ns]']  # <-- master only
```

master
```
>>> for x in news:
...     x[0]
... 
Timestamp('2019-01-07 12:28:45-0800', tz='US/Pacific')
Timestamp('2019-01-07 20:28:45-0800', tz='US/Pacific')
Timestamp('2019-01-07 20:28:45-0800', tz='US/Pacific')
Timestamp('2019-01-07 12:28:45-0800', tz='US/Pacific')
Timestamp('2019-01-07 12:28:45-0800', tz='US/Pacific')
Timestamp('2019-01-07 20:28:45-0800', tz='US/Pacific')

>>> for x in simples:
...     x[0]
... 
Timestamp('2019-01-07 12:28:45-0800', tz='US/Pacific')
Timestamp('2019-01-07 12:28:45-0800', tz='US/Pacific')

>>> for x in inits:
...     x[0]
... 
Timestamp('2019-01-07 12:28:45-0800', tz='US/Pacific')
Timestamp('2019-01-07 12:28:45-0800', tz='US/Pacific')
```

0.23.4
```
>>> for x in news:
...     x[0]
... 
Timestamp('2019-01-07 20:28:45-0800', tz='US/Pacific')
Timestamp('2019-01-07 20:28:45-0800', tz='US/Pacific')
Timestamp('2019-01-07 20:28:45-0800', tz='US/Pacific')
Timestamp('2019-01-07 20:28:45-0800', tz='US/Pacific')
Timestamp('2019-01-07 20:28:45-0800', tz='US/Pacific')
Timestamp('2019-01-07 20:28:45-0800', tz='US/Pacific')
>>> 
>>> for x in simples:
...     x[0]
... 
Timestamp('2019-01-07 12:28:45-0800', tz='US/Pacific')
Timestamp('2019-01-07 12:28:45-0800', tz='US/Pacific')
```

In both 0.23.4 and master, `Timestamp.__new__` treats all i8, M8[ns], and f8 as unix timestamps.

In both 0.23.4 and master, both `DTA.__init__` and `DTI._simple_new` treat both i8 and M8[ns] as unix timestamps.

In 0.23.4 `DTI.__new__` treats all the cases as wall-times.

In master, `DTI.__new__` treats i8, i8+object, and M8[ns]+object as unix timestamps, and the other cases as wall times.

In master, `DTA.__init__` treats both i8 and M8[ns] as unix timestamps.

----
I think there is a consensus that the preferred behavior is for all of these to match Timestamp, right?   So the options for DTI are:

1) revert to 0.23.4 behavior and issue warnings that this behavior is going to change.
2) keep the half-and-half behavior, issue warnings that more is going to change, document that some has changed
3) move all cases to match Timestamp behavior, document the change, possibly issue warnings(?)

And the options for DTA:
a) if option 3 is chosen for DTI, then DTA is fine as-is.
b) otherwise:
    i) change DTA to match DTI, issue equivalent warnings and change them at the same time in the future
    ii) live with them behaving differently.

The question I'd pose to the more senior folks: can we get away with option 3?

**Comment by TomAugspurger** [comment]
I'm having some trouble following the samples in 
https://github.com/pandas-dev/pandas/issues/24559#issuecomment-452075476, but I think that your

> I think there is a consensus that the preferred behavior is for all of these to match Timestamp, right?

is correct.

Can you clarify the difference between your option 1 and 3? Is the difference the handling of i8 values in DTI._simple_new?

**Comment by jbrockmendel** [comment]
> Can you clarify the difference between your option 1 and 3? Is the difference the handling of i8 values in DTI._simple_new?

None of the options I have in mind touch `_simple_new`.

In option 1, `DTI.__new__` behavior is reverted to 0.23.4, where both int64 and datetime64[ns] are treated as wall-times.  This is the maximally-different-from-Timestamp behavior, but the minimal-user-facing-change option.

In option 3, `DTI.__new__` is behavior changed to fully match Timestamp behavior (and `DTA.__init__` behavior, for the subset of inputs that it supports).  This would be the clear choice if backward-compatibility and deprecation cycles weren't a consideration.

**Comment by jorisvandenbossche** [comment]
Thanks for the overview @jbrockmendel 

I also agree that treating integers as unix timestamps would be the preferred behaviour (without the historical baggage) 
Are we then agreeing that the current behaviour in master is the long term behaviour we would like to see? So integers consistently as unix timestamps across constructors, but meaning that ints and ints viewed as M8[ns] will be treated differently. 

(I still find this inconsistency between integers and M8[ns] a bit annoying, but I suppose there is nothing to do about it .. ?)

---

I think I already hinted on my preference before: I would go for option 1 "revert to 0.23.4 behavior and issue deprecation warnings that this behavior is going to change." to deal with the API change properly.  
And for DTA, I would already directly do the future behaviour (what we have now in master).

Given that this was like that before, I would think it should not add too much complexity to the code to still do the deprecation way (I assume many internal shared methods use `_simple_new` anyhow, which is not affected)

> keep the half-and-half behavior, issue warnings that more is going to change, document that some has changed

What more would we want to change?

**Comment by jorisvandenbossche** [comment]
Putting Tom's comment from  here:

https://github.com/pandas-dev/pandas/issues/24559#issuecomment-452075476 has a nice summary.

@jorisvandenbossche prefers 

1. Reverting the change to `DTI.__new__` for passing integers. Add a warning that it'll change in the future.
2. Keep `DTA.__init__`  as it is now (same as Timestamp IIRC)

I think that's my (=Tom's) preference as well, if I understand things correctly. @jbrockmendel do you need / want any help pushing this across the line? I'm busy with there things most of today, but can carve out some time if needed. It'd be good to do the RC today I think.

**Comment by jorisvandenbossche** [comment]
Sorry, I am still missing something here.

I think we agreed somehow that we want "integers -> unix time, datetime64 -> wall time" (which is what DatetimeIndex is doing on master) on the long term ?

On master (compared to 0.23.4), DatetimeIndex changed to treat integers as unix time (instead of wall time), but kept datetime64 as wall time. One reason that was mentioned above was consistency with Timestamp constructor. 
However, Timestamp is treating *both* integers and datetime64 as unix time. So having DatetimeIndex (in the future after deprecation) and DatetimeArray as the "integers -> unix time, datetime64 -> wall time" behaviour, that would still not be consistent with Timestamp:

```
In [74]: ts = pd.Timestamp.now('US/Pacific').round('s') 

In [76]: ts    
Out[76]: Timestamp('2019-01-09 05:52:33-0800', tz='US/Pacific')

In [79]: pd.Timestamp(ts.value, tz='US/Pacific')
Out[79]: Timestamp('2019-01-09 05:52:33-0800', tz='US/Pacific')

In [80]: pd.Timestamp(np.datetime64(ts.value, 'ns'), tz='US/Pacific')
Out[80]: Timestamp('2019-01-09 05:52:33-0800', tz='US/Pacific')     <--- also unix time, not localizing

In [82]: pd.DatetimeIndex([ts.value], tz='US/Pacific')
Out[82]: DatetimeIndex(['2019-01-09 05:52:33-08:00'], dtype='datetime64[ns, US/Pacific]', freq=None)

In [81]: pd.DatetimeIndex([np.datetime64(ts.value, 'ns')], tz='US/Pacific')  <--- localizing wall time
Out[81]: DatetimeIndex(['2019-01-09 13:52:33-08:00'], dtype='datetime64[ns, US/Pacific]', freq=None)
```

**Comment by TomAugspurger** [comment]
Sorry, I misunderstood the timestamp behavior.

**Comment by jorisvandenbossche** [comment]
One nice thing about the current DatetimeArray behaviour (in master it treats both ints and M8 as unix time, i.e. what is stored under the hood): `DatetimeArray(arr)._data == arr` is always true for M8 data, regardless whether a tz-aware dtype is passed or not.

Example with actual example:

```
In [1]: i8 = pd.Timestamp('2000').value

In [2]: arr = np.array([i8], dtype='M8[ns]')      

In [3]: arr                      
Out[3]: array(['2000-01-01T00:00:00.000000000'], dtype='datetime64[ns]')

In [4]: pd.core.arrays.DatetimeArray(arr, dtype=pd.DatetimeTZDtype(tz='CET'))
Out[4]: 
<DatetimeArray>
['2000-01-01 01:00:00+01:00']
Length: 1, dtype: datetime64[ns, CET]

In [5]: pd.core.arrays.DatetimeArray(arr, dtype=pd.DatetimeTZDtype(tz='CET'))._data
Out[5]: array(['2000-01-01T00:00:00.000000000'], dtype='datetime64[ns]')

In [6]: pd.core.arrays.DatetimeArray(arr.view('i8'), dtype=pd.DatetimeTZDtype(tz='CET'))
Out[6]: 
<DatetimeArray>
['2000-01-01 01:00:00+01:00']
Length: 1, dtype: datetime64[ns, CET]

In [7]: pd.core.arrays.DatetimeArray(arr.view('i8'), dtype=pd.DatetimeTZDtype(tz='CET'))._data 
Out[7]: array(['2000-01-01T00:00:00.000000000'], dtype='datetime64[ns]')
```

So the passed array is simply what is stored under the hood, without any work done by DatetimeArray (no localisation). Personally, I also like this property for the simple DatetimeArray constructor. 
It's consistent with Timestamp, but different as `to_datetime` (and `DatetimeIndex`).

**Comment by jorisvandenbossche** [comment]
> Sorry, I misunderstood the timestamp behavior. Feel free to ignore my opinion on this subject :)

No no, it was generally put forward as a reason, not just by you :-)  And I am also only realizing this now

**Comment by jorisvandenbossche** [comment]
OK, turned @jbrockmendel code into a table (and added the result for his open PR as well):

constructor | arg_type | 0.24.3 | master | #24623
-- | -- | -- | -- | --
Timestamp | int64 | unix epoch | unix epoch | unix epoch
Timestamp | datetime64 | unix epoch | unix epoch | unix epoch
DatetimeIndex | int64 | wall time | unix epoch | unix epoch
DatetimeIndex | datetime64 | wall time | wall time | wall time
DatetimeArray | int64 | -- | unix epoch | unix epoch
DatetimeArray | datetime64 | -- | unix epoch | wall time

The code uses:

```
In [1]: ts = pd.Timestamp('2000', tz='US/Pacific') 
   ...: tz = ts.tz

In [2]: ts                          
Out[2]: Timestamp('2000-01-01 00:00:00-0800', tz='US/Pacific')

In [3]: value = ts.value

In [4]: res_wall_time = pd.Timestamp(value).tz_localize('US/Pacific') 
   ...: res_wall_time       
Out[4]: Timestamp('2000-01-01 08:00:00-0800', tz='US/Pacific')

In [5]: res_unix_epoch = pd.Timestamp(value).tz_localize('UTC').tz_convert('US/Pacific') 
   ...: res_unix_epoch                      
Out[5]: Timestamp('2000-01-01 00:00:00-0800', tz='US/Pacific')
```

as set up, and then basically tests `pd.Timestamp(val, tz=tz)`, `pd.DatetimeIndex([val], tz=tz)`, etc for the different cases. 
Full code and results: https://nbviewer.jupyter.org/gist/jorisvandenbossche/37c17bf6f3a67df8decb1d55fdd6def4

Can some-one check if my code and the assumptions on what is wall time and unix epoch makes sense / matches how they understood it ?

**Comment by mroeschke** [comment]
Thanks for this @jorisvandenbossche! Your code looks correct to me.

I am partial to Option 3, but for `datetime64` for `Timestamp` arguments to be interpreted as wall time like the other constructors. I think at minimum it would be ideal to have consistent handling of an argument type across constructors (regardless of high level vs low level).

Agreed that `tz` in the constructor has been ill-defined. I think the PR adding `tz` to `to_datetime` stalled https://github.com/pandas-dev/pandas/issues/13712. I am actually not the biggest fan of adding `tz` in favor of deprecating `utc` there. I would much rather encourage the use of `tz_localize` and `tz_convert`

**Comment by jorisvandenbossche** [comment]
I think my current position (but has been quite variable :-)), is that I am -1 on option 3, as I want unix time for both int/datetime for DatetimeArray. I think then option 2 makes the most sense (so in the future to change DatetimeIndex to treat int as epochs), potentially with aligning Timestamp with that (as you mention)

**Comment by TomAugspurger** [comment]
@mroeschke I added an option 5 to 
https://github.com/pandas-dev/pandas/issues/24559#issuecomment-452779835 that I think is your "Option 3 but for Timestamp to interpret datetime64 as wall time". Can you confirm that I got it right?

**Comment by mroeschke** [comment]
Per the updated table https://github.com/pandas-dev/pandas/issues/24559#issuecomment-452779835; I am partial toward Option 4 or Option 5 on the basis of argument consistency across constructors even if `DatetimeArray` doesn't explicitly have a `tz` argument. I lean a little more towards Option 5 since datetime64 "appears" more wall-time like datetime strings (subjective opinion).

**Comment by TomAugspurger** [comment]
At the dev meeting, we agreed to revert DatetimeIndex's handling of integers back to how it was for 0.23.4.

We haven't reached consensus on the best path going forward.

**Comment by jorisvandenbossche** [comment]
And I added option 6 ... (which is the one I mentioned above) 
So Option 5 is like Option 3 but with aligning Timestamp and DatetimeIndex, and Option 6 is like Option 2 but with aligning Timestamp and DatetimeIndex

**Comment by jbrockmendel** [comment]
@mroeschke can you comment on test_construction_int_rountrip from #21674?  AFAICT reverting to 0.23.4 behavior means breaking this test.

For anyone wondering it also breaks pandas/tests/indexes/datetimes/test_construction.py test_constructor_with_int_tz, test_construction_with_alt_tz_localize in ways that do not look fixable

**Comment by jbrockmendel** [comment]
BTW I'm with @mroeschke in being OK with either Option4 or Option5, with a slight preference toward Option5.

**Comment by mroeschke** [comment]
@jbrockmendel Looks like I grabbed the test from the bottom example of this comment: https://github.com/pandas-dev/pandas/issues/12619#issuecomment-204760329

I believe test_construction_int_rountrip should break with the rollback because it assumes the input int argument is an epoch timestamp.

Should we xfail these tests until we decide the behavior of passing ints into `DTI`?

**Comment by jbrockmendel** [comment]
> Should we xfail these tests until we decide the behavior of passing ints into DTI?

The thing is there are 51 failing tests (fewer if we don't count parametrized tests separately)...

Wait.. everybody wait!  I have good news!

https://github.com/pandas-dev/pandas/pull/21216/files#diff-87b716a95c390fb4d6ffe62b1626ce73R39

We've documented the DTI i8 behavior change already!

**Comment by jbrockmendel** [comment]
Wait... holy cow wait!  We’ve documented the i8 behavior change!  #21216

**Comment by jorisvandenbossche** [comment]
Yes, we documented the change, as we did it deliberately. But we came back on that deciding we want a deprecation instead.

**Comment by jbrockmendel** [comment]
Either way, I'm calling this good enough that I'm going to get lunch instead of banging my head against these test failures.

**Comment by mroeschke** [comment]
I think in theory #21216 should have changed the behavior of for int arrays as well despite me only mentioning Int64Index in the whatsnew.

Looks like I included tests for this case as well: https://github.com/pandas-dev/pandas/pull/21216/files#diff-f368eb7eb824bd248bf961916d6de7eaR492

**Comment by jbrockmendel** [comment]
> I think in theory #21216 should have changed the behavior of for int arrays as well despite me only mentioning Int64Index in the whatsnew.

Yah, the fact that it is tested and documented makes me much more willing to accept the status quo.

> Feel free to post a link to your WIP branch and I can take a look.

@TomAugspurger https://github.com/jbrockmendel/pandas/tree/i8revert2

**Comment by jreback** [comment]
I like option 4 or 5 (slight preference), don't reallly love 6

**Comment by jorisvandenbossche** [comment]
Regarding Option 5, I am -1 on that for the reason of wanting to keep the DatetimeArray simple (doing no work in the base case) and consistent for ints / datetime64. For the same reason that `DatetimeIndex._simple_new` now treats both ints and datetimes as UTC unix epochs (and I see `DatetimeArray` more similar to `DatetimeIndex._simple_new` as to `DatetimeIndex`).

But, I don't really like making DatetimeIndex(datetime64) to be unix epoch instead of wall time (as in Option 4), because that would make it inconsistent with things like `pd.date_range(start_date, periods=N, tz=..)` where the start_date is interpreted as wall time.

**Comment by jorisvandenbossche** [comment]
And in case it might be useful, the experiment I did earlier today: https://github.com/pandas-dev/pandas/compare/master...jorisvandenbossche:i8revert?expand=1 (only changed the DatetimeIndex constructor without fixing any of the consequential failures)

**Comment by TomAugspurger** [comment]
Back to the deprecation warning. What's the expected behavior of `Int64Index.astype('datetime64[ns, tz]')`? That errored on 0.23.4. Should it just be the same as `DatetimeIndex(Index, dtype='datetime64[ns, tz]')`? Should it throw a warning?

**Comment by TomAugspurger** [comment]
One more question: should `tz='UTC'` emit a warning? IIUC, the behavior of

```
pd.DatetimeIndex([1457537600000000000, 1458059600000000000], tz='UTC')
```

will always be the same.

---

Gotta take a break from this until this afternoon. WIP at https://github.com/TomAugspurger/pandas/tree/revert-i8

That's based off @jorisvandenbossche's start. Should be easy to swap in @jbrockmendel's changes to `_from_sequence` if we prefer that to the extra warn parameter.

**Comment by mroeschke** [comment]
I think `Int64Index.astype('datetime64[ns, tz]')` should raise the same warning as `DatetimeIndex(Index, dtype='datetime64[ns, tz]')` as `Index(int64s, dtype='datetime64[ns, tz]')`

I would raise on `UTC` even thought the outcome is the same just to acclimate the user to the fact that passing _any_ tz with ints will change behavior.

**Comment by jorisvandenbossche** [comment]
Nice deprecation warning! (I would maybe only mention something regarding the reason for the change? integers seen as unix epoch which is defined to be UTC? (although its nanoseconds so not exactly unix epochs))

> One more question: should tz='UTC' emit a warning?

Since it stays the same, ideally no, I would say

> What's the expected behavior of Int64Index.astype('datetime64[ns, tz]')

Ideally the new behaviour IMO (although that might make the implementation more complicated?) 
(if we are sure about the new behaviour of course .. )

**Comment by jorisvandenbossche** [comment]
> I would raise on UTC even thought the outcome is the same just to acclimate the user to the fact that passing any tz with ints will change behavior.

Hmm, yes, that might be a good argument (although I think, as a user, I would have the reflex to complain about it: "I already specify UTC!")

**Comment by mroeschke** [comment]
Fair enough. Yeah from a user perspective, it may feel like a false positive warning. My opinion about warning with `UTC` isn't that strong so I'd be fine not emitting a warning in that case.

**Comment by TomAugspurger** [comment]
FWIW, I'm ambivalent about warning in the UTC case. Either way has downsides.

**Comment by jbrockmendel** [comment]
If long term we want option 5, do we need to get the deprecation for Timestamp behavior in for 0.24.0?

**Comment by TomAugspurger** [comment]
Possibly, but not necessarily for the RC (which I'd like to do immediately
following https://github.com/pandas-dev/pandas/pull/24708)

**Comment by jreback** [comment]
anything more on this for 0.24.0?

**Comment by TomAugspurger** [comment]
I think we're good for 0.24.0.

We do need to settle on a desired long-term behavior though.

**Comment by jbrockmendel** [comment]
How close are we to consensus on this?  It looks like @mroeschke, @jreback, and I have expressed preference for options 4 or 5, with a slight preference towards 5.  @jorisvandenbossche has expressed a preference for option 6.

Anyone else want to weigh in?

**Comment by jbrockmendel** [comment]
@TomAugspurger @jorisvandenbossche @mroeschke @jreback I'd like to settle on a desired long-term behavior and start the appropriate deprecation warnings before 0.25.0.  How far are we from consensus?

**Comment by jreback** [comment]
@jbrockmendel can u do. short summary of this and just confirm that 0.24.2
and master have the same behavior

**Comment by TomAugspurger** [comment]
We discussed this on the call. IIRC someone volunteered / was volunteered to summarize, and maybe close this issue? (was I the one to volunteer?)

**Comment by TomAugspurger** [comment]
IIUC, this can be closed since the changes have been made and the deprecation enforced.
